libraries/standard/http/include/http_client.h

```diff
@@ -0,0 +1,357 @@
+#ifndef HTTP_CLIENT_H_
+#define HTTP_CLIENT_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+
+/**
+ * @brief Maximum amount of headers allowed from the server.
+ * 
+ * If the server sends more headers than this configuration, then status code
+ * #HTTP_SECURITY_ALERT_TOO_MANY_HEADERS is returned from #HTTPClient_Send.
+ */
+#ifndef HTTP_MAX_HEADERS_BYTES
+    #define HTTP_MAX_HEADERS_BYTES      2048U
+#endif
+
+/**
+ * @brief The HTTP header "User-Agent" value.
+ * 
+ * The following headerline is automatically written to
+ * #HTTPRequestHeaders_t.pBuffer:
+ * "User-Agent: AWS\r\n"
+ */ 
+#ifndef HTTP_USER_AGENT_VALUE
+    #define HTTP_USER_AGENT_VALUE       "AWS"
+#endif
+
+/**
+ * Supported HTTP request methods.
+ */
+#define HTTP_METHOD_GET                 "GET"  /**< HTTP Method GET. */
+#define HTTP_METHOD_PUT                 "PUT"  /**< HTTP Method GET. */
+#define HTTP_METHOD_POST                "POST" /**< HTTP Method GET. */
+#define HTTP_METHOD_HEAD                "HEAD" /**< HTTP Method GET. */
+
+/**
+ * Flags for #HTTPRequestInfo_t.flags.
+ * These flags control what headers are written or not to the 
+ * #HttpRequestHeaders_t.pBuffer.
+ */
+/**
+ * @brief Set this flag to indicate the request is for a persistent connection.
+ * 
+ * Setting this will cause a "Connection: Keep-Alive" to be written to the
+ * request.
+ * 
+ * Keep in mind some third-party servers do not have to honor persistent
+ * connection requests.
+ */
+#define HTTP_REQUEST_KEEP_ALIVE_FLAG                    0x1U
+/** 
+ * @brief Set this flag to disable automatically writing the Content-Length
+ * header. 
+ */
+#define HTTP_REQUEST_DISABLE_CONTENT_LENGTH_FLAG   0x2U
+
+/**
+ * Flags for #HTTPResponse_t.flags.
+ * These flags are populated in #HTTPResponse_t.flags by the #HTTPClient_Send()
+ * function.
+ */
+ /**
+  * @brief This will be set to true if header "Connection: close" is found.
+  * 
+  * If a "Connection: close" header is present the application should always
+  * close the connection.
+  */
+#define HTTP_RESPONSE_CONNECTION_CLOSE_FLAG             0x1U
+/**
+ * @brief This will be set to true if header "Connection: Keep-Alive" is found.
+ */
+#define HTTP_RESPONSE_CONNECTION_KEEP_ALIVE_FLAG        0x2U
+
+/**
+ * @brief The HTTPNetworkContext is an incomplete type. The application must 
+ * define HTTPNetworkContext to the type of their network context. This context
+ * is passed into the network interface functions.
+ */
+struct HTTPNetworkContext;
+typedef struct HTTPNetworkContext * HTTPNetworkContext_t;
+
+/**
+ * @brief Transport interface for sending data over the network.
+ * 
+ * If the number of bytes written returned is less than bytesToWrite, then 
+ * #HTTPClient_Send will return HTTP_NETWORK_ERROR. If a negative value is 
```

Comment:
Shouldn't `HTTPClient_Send()` return a platform-dependent error code here rather than collapsing all errors into a single value?  At some point this will be built into an application which needs to be debugged, and it would be best if the app developer had all available context needed to debug this without rebuilding/adding logging in lower layers/etc.

Reply:
No because with the interface paradigm, the application owns and implements the transport_write() and transport_read(), so they are aware of the errors that occur there.